packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -511,12 +556,17 @@ export class Stack extends Construct implements ITaggable {
 
   protected synthesize(session: ISynthesisSession): void {
     const builder = session.assembly;
-    const template = `${this.stackName}.template.json`;
+    const template = this.templateFileName;
 
     // write the CloudFormation template as a JSON file
     const outPath = path.join(builder.outdir, template);
     fs.writeFileSync(outPath, JSON.stringify(this._toCloudFormation(), undefined, 2));
 
+    // if this is a nested stack, do not emit it as a cloud assembly artifact (it will be registered as an s3 asset instead)
```

Comment:
Shouldn't this be achieved by overriding the `synthesize` of `NestedStack`?

Reply:
Since I need `parentStack` at the stack level to determine who gets to create cross references, I figured I might as well move most of the nested stack related logic there. Not sure I love it, but I don't think it really matters.